gnome-initial-setup/gis-driver.c

```diff
@@ -127,14 +132,67 @@ check_for_live_boot (gchar **uuid)
   return live_boot;
 }
 
+static void
+udisks_client_new_cb (
+    GObject *source,
+    GAsyncResult *result,
+    gpointer user_data)
+{
+  GisDriver *driver = GIS_DRIVER (user_data);
+  GisDriverPrivate *priv = gis_driver_get_instance_private (driver);
+  UDisksClient *udisks = NULL;
+  GError *error = NULL;
+
+  udisks = udisks_client_new_finish (result, &error);
+  if (udisks != NULL)
+    {
+      GList *blocks = udisks_client_get_block_for_uuid (udisks, priv->live_mode_uuid);
+      GList *l;
+
+      /* If you have the same ISO on a USB and a DVD and you insert both, it is
+       * hard to determine which is booted using UDisks, so assume it's the DVD.
+       */
+      for (l = blocks; l != NULL && !priv->is_live_dvd; l = l->next)
+        {
+          UDisksBlock *block = UDISKS_BLOCK (l->data);
+          UDisksDrive *drive = udisks_client_get_drive_for_block (udisks, block);
+
+          if (drive != NULL && udisks_drive_get_optical (drive))
+            {
+              priv->is_live_dvd = TRUE;
+              g_object_notify_by_pspec (G_OBJECT (driver), obj_props[PROP_LIVE_DVD]);
```

Comment:
It probably does not matter in practice, since the driver is always alive, but I find it a bit surprising that this could potentially notify after the page is "morally" defunct.
Instead you could attach a cancellable to the async operation and return here when that's the case.

Reply:
Fine. I have added a patch to cancel the cancellable when the GisDriver is shut down. (I think you were proposing to not take a ref for the `udisk_client_new()` call but I really dislike this pattern of relying on the precise semantics of whether the returned error is CANCELLED for whether or not the user_data pointer is valid or not.) As explained in the commit message, the process actually never terminates cleanly except in the reformatter path anyway so this code will never run, but okay.

While looking for calls to `g_application_quit`, and while testing this change, I discovered that the error handling path for running `eos-installer` did not work, so I have fixed that. I also fixed some compiler warnings introduced in the most recent rebase.